src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java

```diff
@@ -1189,6 +1195,124 @@ public void commit_NonConflictingDeletesGivenForExisting_ShouldCommitBoth()
+    int current2 = ((IntValue) result2.get().getValue(BALANCE).get()).get();
+    Put put1 = preparePut(0, 0, NAMESPACE, TABLE_1).withValue(new IntValue(BALANCE, current1 + 1));
+    transaction1.put(put1);
+    Put put2 = preparePut(0, 1, NAMESPACE, TABLE_1).withValue(new IntValue(BALANCE, current2 + 1));
+    transaction2.put(put2);
+    transaction1.commit();
+    transaction2.commit();
+
+    // Assert
+    transaction = manager.start();
+    result1 = transaction.get(get1_1);
+    result2 = transaction.get(get2_1);
+    // the results can not be produced by executing the transactions serially
+    assertThat(result1.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 2));
+    assertThat(result2.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 2));
+  }
+
+  public void
+      commit_WriteSkewOnExistingRecordsWithSerializable_OneShouldCommitTheOtherShouldThrowCommitConflictException()
+          throws CommitException, UnknownTransactionStatusException, CrudException {
+    // Arrange
+    List<Put> puts =
+        Arrays.asList(
+            preparePut(0, 0, NAMESPACE, TABLE_1).withValue(new IntValue(BALANCE, 1)),
+            preparePut(0, 1, NAMESPACE, TABLE_1).withValue(new IntValue(BALANCE, 1)));
+    DistributedTransaction transaction = manager.start(Isolation.SERIALIZABLE);
+    transaction.put(puts);
+    transaction.commit();
+
+    // Act
+    DistributedTransaction transaction1 = manager.start(Isolation.SERIALIZABLE);
+    DistributedTransaction transaction2 = manager.start(Isolation.SERIALIZABLE);
+    Get get1_1 = prepareGet(0, 1, NAMESPACE, TABLE_1);
+    Optional<Result> result1 = transaction1.get(get1_1);
+    Get get1_2 = prepareGet(0, 0, NAMESPACE, TABLE_1);
+    transaction1.get(get1_2);
+    int current1 = ((IntValue) result1.get().getValue(BALANCE).get()).get();
+    Get get2_1 = prepareGet(0, 0, NAMESPACE, TABLE_1);
+    Optional<Result> result2 = transaction2.get(get2_1);
+    Get get2_2 = prepareGet(0, 1, NAMESPACE, TABLE_1);
+    transaction2.get(get2_2);
+    int current2 = ((IntValue) result2.get().getValue(BALANCE).get()).get();
+    Put put1 = preparePut(0, 0, NAMESPACE, TABLE_1).withValue(new IntValue(BALANCE, current1 + 1));
+    transaction1.put(put1);
+    Put put2 = preparePut(0, 1, NAMESPACE, TABLE_1).withValue(new IntValue(BALANCE, current2 + 1));
+    transaction2.put(put2);
+    transaction1.commit();
+    Throwable thrown = catchThrowable(transaction2::commit);
+
+    // Assert
+    transaction = manager.start(Isolation.SERIALIZABLE);
+    result1 = transaction.get(get1_1);
+    result2 = transaction.get(get2_1);
+    assertThat(result1.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 1));
+    assertThat(result2.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 2));
+    assertThat(thrown).isInstanceOf(CommitConflictException.class);
+  }
+
+  public void commit_WriteSkewOnNonExistingRecordsWithSerializable_ShouldThrowCommitException()
```

Comment:
This is a test to verify the fix is working as expected.
Since there is no way to detect conflicts for empty records correctly, it will abort when it reads empty record with SERIALIZABLE. It is a false positive abort so there might be a case where a history is serializable without such abort.